python/lsst/daf/butler/queries/tree/_predicate.py

```diff
@@ -155,6 +155,26 @@ def from_bool(cls, value: bool) -> Predicate:
         #
         return cls.model_construct(operands=() if value else ((),))
 
+    @classmethod
+    def from_bool_expression(cls, value: ColumnExpression) -> Predicate:
+        """Construct a predicate that wraps a boolean ColumnExpression, taking
+        on the value of the underlying ColumnExpression.
```

Comment:
I had hoped to never have to allow a boolean `ColumnExpression` to be constructed (going to `Predicate` immediately instead).  But I think I've figured out why you didn't do that, and I agree with it: we'd need `Predicate`-subclass versions of `DimensionFieldReference` and `DatasetFieldReference`, and we'd need to include them in all of the cases where column references of other types already appear, and that'd be a much bigger problem.  Is that right?

If so, I think instead we need to be extra vigilant about putting those column expressions in their predicate wrappers as early as possible, and keeping boolean operators out of those column expressions, because that would subvert all of the boolean-expression normalization and reasoning we do.  I suspect typing is already going to do prevent problems pretty effectively, since we don't have any boolean-operator expressions, but we could lock it down a bit further by declaring that `value` here (and in `BooleanWrapper`) must be `ColumnReference`, not merely `ColumnExpression`.

Reply:
My thinking wasn't actually that involved, I just observed that:
* Boolean ColumnExpressions already existed, and it didn't occur to me to go out of my way to create a separate representation for boolean SQL columns than the one used for int or string.
* Most of the other predicates worked by wrapping one or more column expressions with a little bit of information  specifying how they convert to booleans -- and "no-op" seemed like a reasonable addition to the set of conversions.

So I just went with the grain of what was already there and it slotted in easily with minimal code.

I do share your concern about this problem potentially coming back since any place that references a boolean column will need to do the same thing, but that's sort of inherent to treating booleans as a special case.  Changing the type is a good suggestion, I'll do that.

I suspect handling `boolean_column = NULL` is going to get ugly though so I may need to figure something else out -- or just not support it, or special case it to only work for a single boolean column and not arbitrary boolean expressions.